torch/csrc/jit/script/init.cpp

```diff
@@ -529,8 +529,38 @@ void initJitScriptBindings(PyObject* module) {
           "_register_attribute",
           [](Module& self, std::string name, TypePtr type, py::object value) {
             auto unshaped = unshapedType(type);
-            self.register_attribute(
-                name, unshaped, toIValue(std::move(value), type));
+            if (type->cast<InterfaceType>()) {
+              // When registering an object to a interface, we turn the value
+              // into the compiled TorchScript class and check if it conform
+              // with the interface or not, then register the attribute to our
+              // slots.
+              py::str qualified_name =
+                  py::module::import("torch.jit")
+                      .attr("_qualified_name")(value.get_type());
+              auto pyCu = get_python_cu();
+              const auto classType =
+                  pyCu->get_class(c10::QualifiedName(qualified_name));
+              if (!classType) {
+                throw std::runtime_error(c10::str(
+                    "registering the attribute ",
+                    name,
+                    " as an interface fails because the value is not "
+                    "a TorchScript compatible type, did you forget to",
+                    "turn it into a user defined TorchScript class?"));
+              }
+              if (!classType->isSubtypeOf(type)) {
```

Comment:
This check should not be necessary. It should be handled for all types in register_attribute. Furthermore (if it isn't  already), it should be using isSubtypeExt so that it provides more details about why something doesn't match an interface.

Reply:
Actually this check should not be necessary because we should just error in toIValue if the interface type is not compatible with the classes type.

Reply:
Yeah I will put the `isSubtypeOf` check logic and error out in the `toIValue` instead of here.